`TestHSMMigrate` flakiness

## Failure

#### Link(s) to logs

- https://console.cloud.google.com/cloud-build/builds;region=us-west1/c8063521-9f7a-4663-82ed-01266d39a199?project=ci-account

#### Relevant snippet

```
        	Messages:   	error while closing auth2 during test cleanup
        	Test:       	TestHSMMigrate
        	            	context canceled
        	Error:      	Received unexpected error:
        	            				hsm_test.go:826
        	            				hsm_test.go:811
        	            				testing.go:864
        	            				panic.go:500
        	            				testing.go:1433
        	            				testing.go:1203
        	            				testing.go:1034
        	            				hsm_test.go:76
        	            				testing.go:864
        	            				panic.go:500
        	            				testing.go:1187
        	            				testing.go:1203
        	            				testing.go:1034
        	Error Trace:	hsm_test.go:76
    hsm_test.go:76: 
```

https://github.com/gravitational/teleport/actions/runs/5976245336/job/16215172101

```
2023-08-25T14:05:11.9773890Z     hsm_test.go:78: 
2023-08-25T14:05:11.9774014Z         	Error Trace:	/__w/teleport/teleport/integration/hsm/hsm_test.go:78
2023-08-25T14:05:11.9774115Z         	            				/opt/go/src/testing/testing.go:1150
2023-08-25T14:05:11.9774210Z         	            				/opt/go/src/testing/testing.go:1328
2023-08-25T14:05:11.9774309Z         	            				/opt/go/src/testing/testing.go:1312
2023-08-25T14:05:11.9774406Z         	            				/opt/go/src/runtime/panic.go:540
2023-08-25T14:05:11.9774494Z         	            				/opt/go/src/testing/testing.go:980
2023-08-25T14:05:11.9774610Z         	            				/__w/teleport/teleport/integration/hsm/hsm_test.go:78
2023-08-25T14:05:11.9774708Z         	            				/opt/go/src/testing/testing.go:1150
2023-08-25T14:05:11.9774801Z         	            				/opt/go/src/testing/testing.go:1328
2023-08-25T14:05:11.9774894Z         	            				/opt/go/src/testing/testing.go:1570
2023-08-25T14:05:11.9774989Z         	            				/opt/go/src/runtime/panic.go:540
2023-08-25T14:05:11.9775084Z         	            				/opt/go/src/testing/testing.go:980
2023-08-25T14:05:11.9775189Z         	            				/__w/teleport/teleport/integration/hsm/hsm_test.go:820
2023-08-25T14:05:11.9775298Z         	            				/__w/teleport/teleport/integration/hsm/hsm_test.go:879
2023-08-25T14:05:11.9775396Z         	Error:      	Received unexpected error:
2023-08-25T14:05:11.9775514Z         	            	context canceled, context canceled, context canceled
2023-08-25T14:05:11.9775591Z         	Test:       	TestHSMMigrate
2023-08-25T14:05:11.9775708Z         	Messages:   	error while closing auth1 during test cleanup
...
2023-08-25T14:05:11.9779186Z --- FAIL: TestHSMMigrate (480.08s)
```

--- a/integration/hsm/helpers.go
+++ b/integration/hsm/helpers.go
@@ -236,6 +236,7 @@ func newAuthConfig(t *testing.T, log utils.Logger) *servicecfg.Config {
 	config.Log = log
 	config.InstanceMetadataClient = cloud.NewDisabledIMDSClient()
 	config.MaxRetryPeriod = 25 * time.Millisecond
+	config.PollingPeriod = 2 * time.Second
 
 	config.Auth.Enabled = true
 	config.Auth.NoAudit = true
@@ -278,6 +279,7 @@ func newProxyConfig(t *testing.T, authAddr utils.NetAddr, log utils.Logger) *ser
 	config.Log = log
 	config.InstanceMetadataClient = cloud.NewDisabledIMDSClient()
 	config.MaxRetryPeriod = 25 * time.Millisecond
+	config.PollingPeriod = 2 * time.Second
 
 	config.Proxy.Enabled = true
 	config.Proxy.DisableWebInterface = true

--- a/integration/hsm/helpers.go
+++ b/integration/hsm/helpers.go
@@ -191,6 +191,12 @@ func (t *teleportService) authAddr(testingT *testing.T) utils.NetAddr {
 	return *addr
 }
 
+func (t *teleportService) authAddrString(testingT *testing.T) string {
+	addr, err := t.process.AuthAddr()
+	require.NoError(testingT, err)
+	return addr.String()
+}
+
 type teleportServices []*teleportService
 
 func (s teleportServices) forEach(f func(t *teleportService) error) error {

--- a/integration/hsm/helpers.go
+++ b/integration/hsm/helpers.go
@@ -40,12 +40,13 @@ import (
 // up. Methods are not meant to be called concurrently on the same receiver and
 // are not generally thread safe.
 type teleportService struct {
-	name           string
-	log            utils.Logger
-	config         *servicecfg.Config
-	process        *service.TeleportProcess
-	serviceChannel chan *service.TeleportProcess
-	errorChannel   chan error
+	name              string
+	log               utils.Logger
+	config            *servicecfg.Config
+	process           *service.TeleportProcess
+	processGeneration int
+	serviceChannel    chan *service.TeleportProcess
+	errorChannel      chan error
 }
 
 func newTeleportService(t *testing.T, config *servicecfg.Config, name string) *teleportService {
@@ -77,73 +78,76 @@ func (t *teleportService) start(ctx context.Context) error {
 	// receive all new processes after restarts and write them to a goroutine.
 	go func() {
 		t.errorChannel <- service.Run(ctx, *t.config, func(cfg *servicecfg.Config) (service.Process, error) {
-			t.log.Debugf("(Re)starting %s", t.name)
+			t.log.Debugf("%s gen %d: starting next process generation (gen %d)", t.name, t.processGeneration, t.processGeneration+1)
 			svc, err := service.NewTeleport(cfg)
 			if err == nil {
-				t.log.Debugf("Started %s, writing to serviceChannel", t.name)
+				t.log.Debugf("%s gen %d: started, writing to serviceChannel", t.name, t.processGeneration+1)
 				t.serviceChannel <- svc
 			}
 			return svc, trace.Wrap(err)
 		})
 	}()
-	t.log.Debugf("Waiting for %s to start", t.name)
+	t.log.Debugf("%s gen 1: waiting for first start", t.name)
 	if err := t.waitForNewProcess(ctx); err != nil {
 		return trace.Wrap(err)
 	}
-	t.log.Debugf("%s started, waiting for it to be ready", t.name)
+	t.log.Debugf("%s gen 1: started, waiting for it to be ready", t.name)
 	return t.waitForReady(ctx)
 }
 
 func (t *teleportService) waitForNewProcess(ctx context.Context) error {
 	select {
 	case t.process = <-t.serviceChannel:
-		t.log.Debugf("received new process for %s from serviceChannel", t.name)
+		t.processGeneration += 1
+		t.log.Debugf("%s gen %d: received new process from serviceChannel", t.name, t.processGeneration)
 	case err := <-t.errorChannel:
 		return trace.Wrap(err)
 	case <-ctx.Done():
-		return trace.Wrap(ctx.Err(), "timed out waiting for %s to restart", t.name)
+		return trace.Wrap(ctx.Err(), "%s gen %d: timed out waiting for restart", t.name, t.processGeneration)
 	}
 	return nil
 }
 
 func (t *teleportService) waitForReady(ctx context.Context) error {
-	t.log.Debugf("Waiting for %s to be ready", t.name)
+	t.log.Debugf("%s gen %d: waiting for TeleportReadyEvent", t.name, t.processGeneration)
 	if _, err := t.process.WaitForEvent(ctx, service.TeleportReadyEvent); err != nil {
-		return trace.Wrap(err, "timed out waiting for %s to be ready", t.name)
+		return trace.Wrap(err, "timed out waiting for %s gen %d to be ready", t.name, t.processGeneration)
 	}
-	// If this is an Auth servier, also wait for AuthIdentityEvent so that we
+	t.log.Debugf("%s gen %d: got TeleportReadyEvent", t.name, t.processGeneration)
+	// If this is an Auth server, also wait for AuthIdentityEvent so that we
 	// can safely read the admin credentials and create a test client.
 	if t.process.GetAuthServer() != nil {
+		t.log.Debugf("%s gen %d: waiting for AuthIdentityEvent", t.name, t.processGeneration)
 		if _, err := t.process.WaitForEvent(ctx, service.AuthIdentityEvent); err != nil {
-			return trace.Wrap(err, "timed out waiting for %s auth identity event", t.name)
+			return trace.Wrap(err, "%s gen %d: timed out waiting AuthIdentityEvent", t.name, t.processGeneration)
 		}
-		t.log.Debugf("%s is ready", t.name)
+		t.log.Debugf("%s gen %d: got AuthIdentityEvent", t.name, t.processGeneration)
 	}
 	return nil
 }
 
 func (t *teleportService) waitForRestart(ctx context.Context) error {
-	t.log.Debugf("Waiting for %s to restart", t.name)
+	t.log.Debugf("%s gen %d: waiting for restart", t.name, t.processGeneration)
 	if err := t.waitForNewProcess(ctx); err != nil {
 		return trace.Wrap(err)
 	}
-	t.log.Debugf("%s restarted, waiting for new process to be ready", t.name)
+	t.log.Debugf("%s gen %d: restarted, waiting for new process (gen %d) to be ready", t.name, t.processGeneration-1, t.processGeneration)
 	return trace.Wrap(t.waitForReady(ctx))
 }
 
 func (t *teleportService) waitForShutdown(ctx context.Context) error {
-	t.log.Debugf("Waiting for %s to shut down", t.name)
+	t.log.Debugf("%s gen %d: waiting for shutdown", t.name, t.processGeneration)
 	select {
 	case err := <-t.errorChannel:
 		t.process = nil
 		return trace.Wrap(err)
 	case <-ctx.Done():
-		return trace.Wrap(ctx.Err(), "timed out waiting for %s to shut down", t.name)
+		return trace.Wrap(ctx.Err(), "%s gen %d: timed out waiting for shutdown", t.name, t.processGeneration)
 	}
 }
 
 func (t *teleportService) waitForLocalAdditionalKeys(ctx context.Context) error {
-	t.log.Debugf("Waiting for %s to have local additional keys", t.name)
+	t.log.Debugf("%s gen %d: waiting for local additional keys", t.name, t.processGeneration)
 	clusterName, err := t.process.GetAuthServer().GetClusterName()
 	if err != nil {
 		return trace.Wrap(err)
@@ -152,7 +156,7 @@ func (t *teleportService) waitForLocalAdditionalKeys(ctx context.Context) error 
 	for {
 		select {
 		case <-ctx.Done():
-			return trace.Wrap(ctx.Err(), "timed out waiting for %s to have local additional keys", t.name)
+			return trace.Wrap(ctx.Err(), "%s gen %d: timed out waiting for local additional keys", t.name, t.processGeneration)
 		case <-time.After(250 * time.Millisecond):
 		}
 		ca, err := t.process.GetAuthServer().GetCertAuthority(ctx, hostCAID, true)
@@ -167,16 +171,16 @@ func (t *teleportService) waitForLocalAdditionalKeys(ctx context.Context) error 
 			break
 		}
 	}
-	t.log.Debugf("%s has local additional keys", t.name)
+	t.log.Debugf("%s gen %d has local additional keys", t.name, t.processGeneration)
 	return nil
 }
 
 func (t *teleportService) waitForPhaseChange(ctx context.Context) error {
-	t.log.Debugf("Waiting for %s to change phase", t.name)
+	t.log.Debugf("%s gen %d: waiting for phase change", t.name, t.processGeneration)
 	if _, err := t.process.WaitForEvent(ctx, service.TeleportPhaseChangeEvent); err != nil {
-		return trace.Wrap(err, "timed out waiting for %s to change phase", t.name)
+		return trace.Wrap(err, "%s gen %d: timed out waiting for phase change", t.name, t.processGeneration)
 	}
-	t.log.Debugf("%s changed phase", t.name)
+	t.log.Debugf("%s gen %d: changed phase", t.name, t.processGeneration)
 	return nil
 }
 

--- a/integration/hsm/hsm_test.go
+++ b/integration/hsm/hsm_test.go
@@ -186,10 +186,6 @@ func TestHSMRotation(t *testing.T) {
 
 // Tests multiple CA rotations and rollbacks with 2 HSM auth servers in an HA configuration
 func TestHSMDualAuthRotation(t *testing.T) {
-	// TODO(nklaassen): fix this test and re-enable it.
-	// https://github.com/gravitational/teleport/issues/20217
-	t.Skip("TestHSMDualAuthRotation is temporarily disabled due to flakiness")
-
 	requireHSMAvailable(t)
 	requireETCDAvailable(t)
 

--- a/integration/hsm/hsm_test.go
+++ b/integration/hsm/hsm_test.go
@@ -84,7 +84,7 @@ func etcdBackendConfig(t *testing.T) *backend.Config {
 	t.Cleanup(func() {
 		bk, err := etcdbk.New(context.Background(), cfg.Params)
 		require.NoError(t, err)
-		require.NoError(t, bk.DeleteRange(context.Background(), []byte(prefix),
+		require.NoError(t, bk.DeleteRange(context.Background(), append([]byte("/"), []byte(prefix)...),
 			backend.RangeEnd([]byte(prefix))),
 			"failed to clean up etcd backend")
 	})

--- a/integration/hsm/hsm_test.go
+++ b/integration/hsm/hsm_test.go
@@ -23,7 +23,6 @@ import (
 	"time"
 
 	"github.com/google/uuid"
-	"github.com/gravitational/trace"
 	"github.com/stretchr/testify/require"
 
 	"github.com/gravitational/teleport"
@@ -184,6 +183,44 @@ func TestHSMRotation(t *testing.T) {
 	require.NoError(t, allServices.waitForRestart(ctx))
 }
 
+func getAdminClient(t *testing.T, authDataDir string, authAddr string) *auth.Client {
+	identity, err := auth.ReadLocalIdentity(
+		filepath.Join(authDataDir, teleport.ComponentProcess),
+		auth.IdentityID{Role: types.RoleAdmin})
+	require.NoError(t, err)
+
+	tlsConfig, err := identity.TLSConfig(nil /*cipherSuites*/)
+	require.NoError(t, err)
+
+	clt, err := auth.NewClient(client.Config{
+		Addrs: []string{authAddr},
+		Credentials: []client.Credentials{
+			client.LoadTLS(tlsConfig),
+		},
+		CircuitBreakerConfig: breaker.NoopBreakerConfig(),
+	})
+	require.NoError(t, err)
+
+	return clt
+}
+
+func testAdminClient(t *testing.T, authDataDir string, authAddr string) {
+	var errs []error
+	require.Eventually(t, func() bool {
+		clt := getAdminClient(t, authDataDir, authAddr)
+		// Make sure it succeeds twice in a row, we might be hitting a load
+		// balancer in front of two auths, this gives a better chance of testing
+		// both
+		for i := 0; i < 2; i++ {
+			if _, err := clt.GetClusterName(); err != nil {
+				errs = append(errs, err)
+				return false
+			}
+		}
+		return true
+	}, 10*time.Second, time.Second, "admin client failed test call to GetClusterName, observed errors: %v", errs)
+}
+
 // Tests multiple CA rotations and rollbacks with 2 HSM auth servers in an HA configuration
 func TestHSMDualAuthRotation(t *testing.T) {
 	requireHSMAvailable(t)
@@ -238,31 +275,10 @@ func TestHSMDualAuthRotation(t *testing.T) {
 	authServices = append(authServices, auth2)
 	allServices = append(allServices, auth2)
 
-	// make sure the admin identity used by tctl works
-	getAdminClient := func() *auth.Client {
-		identity, err := auth.ReadLocalIdentity(
-			filepath.Join(auth2Config.DataDir, teleport.ComponentProcess),
-			auth.IdentityID{Role: types.RoleAdmin, HostUUID: auth2Config.HostUUID})
-		require.NoError(t, err)
-		tlsConfig, err := identity.TLSConfig(nil)
-		require.NoError(t, err)
-		authAddrs := []utils.NetAddr{auth2.authAddr(t)}
-		clt, err := auth.NewClient(client.Config{
-			Addrs: utils.NetAddrsToStrings(authAddrs),
-			Credentials: []client.Credentials{
-				client.LoadTLS(tlsConfig),
-			},
-			CircuitBreakerConfig: breaker.NoopBreakerConfig(),
-		})
-		require.NoError(t, err)
-		return clt
+	testAuth2Client := func(t *testing.T) {
+		testAdminClient(t, auth2Config.DataDir, auth2.authAddrString(t))
 	}
-	testClient := func(clt *auth.Client) error {
-		_, err = clt.GetClusterName()
-		return trace.Wrap(err)
-	}
-	clt := getAdminClient()
-	require.NoError(t, testClient(clt))
+	testAuth2Client(t)
 
 	stages := []struct {
 		targetPhase string
@@ -273,32 +289,28 @@ func TestHSMDualAuthRotation(t *testing.T) {
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForPhaseChange(ctx))
 				require.NoError(t, authServices.waitForLocalAdditionalKeys(ctx))
-				clt = getAdminClient()
-				require.NoError(t, testClient(clt))
+				testAuth2Client(t)
 			},
 		},
 		{
 			targetPhase: types.RotationPhaseUpdateClients,
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForRestart(ctx))
-				clt = getAdminClient()
-				require.NoError(t, testClient(clt))
+				testAuth2Client(t)
 			},
 		},
 		{
 			targetPhase: types.RotationPhaseUpdateServers,
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForRestart(ctx))
-				clt = getAdminClient()
-				require.NoError(t, testClient(clt))
+				testAuth2Client(t)
 			},
 		},
 		{
 			targetPhase: types.RotationPhaseStandby,
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForRestart(ctx))
-				clt = getAdminClient()
-				require.NoError(t, testClient(clt))
+				testAuth2Client(t)
 			},
 		},
 	}
@@ -317,32 +329,10 @@ func TestHSMDualAuthRotation(t *testing.T) {
 	// Safe to send traffic to new auth server now that a full rotation has been completed.
 	lb.AddBackend(auth2.authAddr(t))
 
-	// load balanced client shoud work with either backend
-	getAdminClient = func() *auth.Client {
-		identity, err := auth.ReadLocalIdentity(
-			filepath.Join(auth2Config.DataDir, teleport.ComponentProcess),
-			auth.IdentityID{Role: types.RoleAdmin, HostUUID: auth2Config.HostUUID})
-		require.NoError(t, err)
-		tlsConfig, err := identity.TLSConfig(nil)
-		require.NoError(t, err)
-		authAddrs := []string{lb.Addr().String()}
-		clt, err := auth.NewClient(client.Config{
-			Addrs: authAddrs,
-			Credentials: []client.Credentials{
-				client.LoadTLS(tlsConfig),
-			},
-			CircuitBreakerConfig: breaker.NoopBreakerConfig(),
-		})
-		require.NoError(t, err)
-		return clt
+	testLoadBalancedClient := func(t *testing.T) {
+		testAdminClient(t, auth2Config.DataDir, lb.Addr().String())
 	}
-	testClient = func(clt *auth.Client) error {
-		_, err1 := clt.GetClusterName()
-		_, err2 := clt.GetClusterName()
-		return trace.NewAggregate(err1, err2)
-	}
-	clt = getAdminClient()
-	require.NoError(t, testClient(clt))
+	testLoadBalancedClient(t)
 
 	// Do another full rotation from the new auth server
 	for _, stage := range stages {
@@ -365,24 +355,21 @@ func TestHSMDualAuthRotation(t *testing.T) {
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForPhaseChange(ctx))
 				require.NoError(t, authServices.waitForLocalAdditionalKeys(ctx))
-				clt := getAdminClient()
-				require.NoError(t, testClient(clt))
+				testLoadBalancedClient(t)
 			},
 		},
 		{
 			targetPhase: types.RotationPhaseRollback,
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForRestart(ctx))
-				clt := getAdminClient()
-				require.NoError(t, testClient(clt))
+				testLoadBalancedClient(t)
 			},
 		},
 		{
 			targetPhase: types.RotationPhaseStandby,
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForRestart(ctx))
-				clt := getAdminClient()
-				require.NoError(t, testClient(clt))
+				testLoadBalancedClient(t)
 			},
 		},
 		{
@@ -390,32 +377,28 @@ func TestHSMDualAuthRotation(t *testing.T) {
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForPhaseChange(ctx))
 				require.NoError(t, authServices.waitForLocalAdditionalKeys(ctx))
-				clt := getAdminClient()
-				require.NoError(t, testClient(clt))
+				testLoadBalancedClient(t)
 			},
 		},
 		{
 			targetPhase: types.RotationPhaseUpdateClients,
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForRestart(ctx))
-				clt := getAdminClient()
-				require.NoError(t, testClient(clt))
+				testLoadBalancedClient(t)
 			},
 		},
 		{
 			targetPhase: types.RotationPhaseRollback,
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForRestart(ctx))
-				clt := getAdminClient()
-				require.NoError(t, testClient(clt))
+				testLoadBalancedClient(t)
 			},
 		},
 		{
 			targetPhase: types.RotationPhaseStandby,
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForRestart(ctx))
-				clt := getAdminClient()
-				require.NoError(t, testClient(clt))
+				testLoadBalancedClient(t)
 			},
 		},
 		{
@@ -423,40 +406,35 @@ func TestHSMDualAuthRotation(t *testing.T) {
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForPhaseChange(ctx))
 				require.NoError(t, authServices.waitForLocalAdditionalKeys(ctx))
-				clt := getAdminClient()
-				require.NoError(t, testClient(clt))
+				testLoadBalancedClient(t)
 			},
 		},
 		{
 			targetPhase: types.RotationPhaseUpdateClients,
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForRestart(ctx))
-				clt := getAdminClient()
-				require.NoError(t, testClient(clt))
+				testLoadBalancedClient(t)
 			},
 		},
 		{
 			targetPhase: types.RotationPhaseUpdateServers,
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForRestart(ctx))
-				clt := getAdminClient()
-				require.NoError(t, testClient(clt))
+				testLoadBalancedClient(t)
 			},
 		},
 		{
 			targetPhase: types.RotationPhaseRollback,
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForRestart(ctx))
-				clt := getAdminClient()
-				require.NoError(t, testClient(clt))
+				testLoadBalancedClient(t)
 			},
 		},
 		{
 			targetPhase: types.RotationPhaseStandby,
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForRestart(ctx))
-				clt := getAdminClient()
-				require.NoError(t, testClient(clt))
+				testLoadBalancedClient(t)
 			},
 		},
 	}
@@ -513,32 +491,10 @@ func TestHSMMigrate(t *testing.T) {
 	proxy := newTeleportService(t, proxyConfig, "proxy")
 	require.NoError(t, proxy.start(ctx))
 
-	// make sure the admin identity used by tctl works
-	getAdminClient := func() *auth.Client {
-		identity, err := auth.ReadLocalIdentity(
-			filepath.Join(auth2Config.DataDir, teleport.ComponentProcess),
-			auth.IdentityID{Role: types.RoleAdmin, HostUUID: auth2Config.HostUUID})
-		require.NoError(t, err)
-		tlsConfig, err := identity.TLSConfig(nil)
-		require.NoError(t, err)
-		authAddrs := []utils.NetAddr{auth2.authAddr(t)}
-		clt, err := auth.NewClient(client.Config{
-			Addrs: utils.NetAddrsToStrings(authAddrs),
-			Credentials: []client.Credentials{
-				client.LoadTLS(tlsConfig),
-			},
-			CircuitBreakerConfig: breaker.NoopBreakerConfig(),
-		})
-		require.NoError(t, err)
-		return clt
+	testClient := func(t *testing.T) {
+		testAdminClient(t, auth2Config.DataDir, auth2.authAddrString(t))
 	}
-	testClient := func(clt *auth.Client) error {
-		_, err1 := clt.GetClusterName()
-		_, err2 := clt.GetClusterName()
-		return trace.NewAggregate(err1, err2)
-	}
-	clt := getAdminClient()
-	require.NoError(t, testClient(clt))
+	testClient(t)
 
 	// Phase 1: migrate auth1 to HSM
 	lb.RemoveBackend(auth1.authAddr(t))
@@ -548,8 +504,7 @@ func TestHSMMigrate(t *testing.T) {
 	auth1 = newTeleportService(t, auth1Config, "auth1")
 	require.NoError(t, auth1.start(ctx))
 
-	clt = getAdminClient()
-	require.NoError(t, testClient(clt))
+	testClient(t)
 
 	authServices := teleportServices{auth1, auth2}
 	allServices := teleportServices{auth1, auth2, proxy}
@@ -563,32 +518,28 @@ func TestHSMMigrate(t *testing.T) {
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForPhaseChange(ctx))
 				require.NoError(t, authServices.waitForLocalAdditionalKeys(ctx))
-				clt := getAdminClient()
-				require.NoError(t, testClient(clt))
+				testClient(t)
 			},
 		},
 		{
 			targetPhase: types.RotationPhaseUpdateClients,
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForRestart(ctx))
-				clt = getAdminClient()
-				require.NoError(t, testClient(clt))
+				testClient(t)
 			},
 		},
 		{
 			targetPhase: types.RotationPhaseUpdateServers,
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForRestart(ctx))
-				clt = getAdminClient()
-				require.NoError(t, testClient(clt))
+				testClient(t)
 			},
 		},
 		{
 			targetPhase: types.RotationPhaseStandby,
 			verify: func(t *testing.T) {
 				require.NoError(t, allServices.waitForRestart(ctx))
-				clt = getAdminClient()
-				require.NoError(t, testClient(clt))
+				testClient(t)
 			},
 		},
 	}
@@ -618,8 +569,7 @@ func TestHSMMigrate(t *testing.T) {
 	authServices = teleportServices{auth1, auth2}
 	allServices = teleportServices{auth1, auth2, proxy}
 
-	clt = getAdminClient()
-	require.NoError(t, testClient(clt))
+	testClient(t)
 
 	// do a full rotation
 	for _, stage := range stages {
@@ -634,5 +584,5 @@ func TestHSMMigrate(t *testing.T) {
 
 	// Safe to send traffic to new auth2 again
 	lb.AddBackend(auth2.authAddr(t))
-	require.NoError(t, testClient(clt))
+	testClient(t)
 }

--- a/integration/hsm/hsm_test.go
+++ b/integration/hsm/hsm_test.go
@@ -24,6 +24,7 @@ import (
 	"time"
 
 	"github.com/google/uuid"
+	"github.com/gravitational/trace"
 	"github.com/stretchr/testify/require"
 
 	"github.com/gravitational/teleport"
@@ -38,7 +39,6 @@ import (
 	"github.com/gravitational/teleport/lib/modules"
 	"github.com/gravitational/teleport/lib/service/servicecfg"
 	"github.com/gravitational/teleport/lib/utils"
-	"github.com/gravitational/trace"
 )
 
 func TestMain(m *testing.M) {

--- a/integration/hsm/hsm_test.go
+++ b/integration/hsm/hsm_test.go
@@ -16,6 +16,7 @@ package hsm
 
 import (
 	"context"
+	"errors"
 	"net"
 	"os"
 	"path/filepath"
@@ -37,6 +38,7 @@ import (
 	"github.com/gravitational/teleport/lib/modules"
 	"github.com/gravitational/teleport/lib/service/servicecfg"
 	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
 )
 
 func TestMain(m *testing.M) {
@@ -83,8 +85,12 @@ func etcdBackendConfig(t *testing.T) *backend.Config {
 	t.Cleanup(func() {
 		bk, err := etcdbk.New(context.Background(), cfg.Params)
 		require.NoError(t, err)
-		require.NoError(t, bk.DeleteRange(context.Background(), append([]byte("/"), []byte(prefix)...),
-			backend.RangeEnd([]byte(prefix))),
+
+		// Based on [backend.Sanitizer] these define the possible range that
+		// needs to be cleaned up at the end of the test.
+		firstPossibleKey := []byte("+")
+		lastPossibleKey := backend.RangeEnd([]byte("z"))
+		require.NoError(t, bk.DeleteRange(context.Background(), firstPossibleKey, lastPossibleKey),
 			"failed to clean up etcd backend")
 	})
 	return cfg
@@ -183,14 +189,18 @@ func TestHSMRotation(t *testing.T) {
 	require.NoError(t, allServices.waitForRestart(ctx))
 }
 
-func getAdminClient(t *testing.T, authDataDir string, authAddr string) *auth.Client {
+func getAdminClient(authDataDir string, authAddr string) (*auth.Client, error) {
 	identity, err := auth.ReadLocalIdentity(
 		filepath.Join(authDataDir, teleport.ComponentProcess),
 		auth.IdentityID{Role: types.RoleAdmin})
-	require.NoError(t, err)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
 
 	tlsConfig, err := identity.TLSConfig(nil /*cipherSuites*/)
-	require.NoError(t, err)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
 
 	clt, err := auth.NewClient(client.Config{
 		Addrs: []string{authAddr},
@@ -199,15 +209,17 @@ func getAdminClient(t *testing.T, authDataDir string, authAddr string) *auth.Cli
 		},
 		CircuitBreakerConfig: breaker.NoopBreakerConfig(),
 	})
-	require.NoError(t, err)
-
-	return clt
+	return clt, trace.Wrap(err)
 }
 
 func testAdminClient(t *testing.T, authDataDir string, authAddr string) {
 	var errs []error
 	require.Eventually(t, func() bool {
-		clt := getAdminClient(t, authDataDir, authAddr)
+		clt, err := getAdminClient(authDataDir, authAddr)
+		if err != nil {
+			errs = append(errs, err)
+			return false
+		}
 		// Make sure it succeeds twice in a row, we might be hitting a load
 		// balancer in front of two auths, this gives a better chance of testing
 		// both
@@ -218,7 +230,7 @@ func testAdminClient(t *testing.T, authDataDir string, authAddr string) {
 			}
 		}
 		return true
-	}, 10*time.Second, time.Second, "admin client failed test call to GetClusterName, observed errors: %v", errs)
+	}, 10*time.Second, time.Second, "admin client failed test call to GetClusterName, observed errors: %v", errors.Join(errs...))
 }
 
 // Tests multiple CA rotations and rollbacks with 2 HSM auth servers in an HA configuration

--- a/integration/hsm/hsm_test.go
+++ b/integration/hsm/hsm_test.go
@@ -16,7 +16,6 @@ package hsm
 
 import (
 	"context"
-	"errors"
 	"net"
 	"os"
 	"path/filepath"
@@ -25,6 +24,7 @@ import (
 
 	"github.com/google/uuid"
 	"github.com/gravitational/trace"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
 	"github.com/gravitational/teleport"
@@ -213,24 +213,20 @@ func getAdminClient(authDataDir string, authAddr string) (*auth.Client, error) {
 }
 
 func testAdminClient(t *testing.T, authDataDir string, authAddr string) {
-	var errs []error
-	require.Eventually(t, func() bool {
+	require.EventuallyWithT(t, func(t *assert.CollectT) {
 		clt, err := getAdminClient(authDataDir, authAddr)
+		assert.NoError(t, err)
 		if err != nil {
-			errs = append(errs, err)
-			return false
+			return
 		}
 		// Make sure it succeeds twice in a row, we might be hitting a load
 		// balancer in front of two auths, this gives a better chance of testing
 		// both
 		for i := 0; i < 2; i++ {
-			if _, err := clt.GetClusterName(); err != nil {
-				errs = append(errs, err)
-				return false
-			}
+			_, err := clt.GetClusterName()
+			assert.NoError(t, err)
 		}
-		return true
-	}, 10*time.Second, time.Second, "admin client failed test call to GetClusterName, observed errors: %v", errors.Join(errs...))
+	}, 10*time.Second, time.Second, "admin client failed test call to GetClusterName")
 }
 
 // Tests multiple CA rotations and rollbacks with 2 HSM auth servers in an HA configuration

--- a/lib/service/service.go
+++ b/lib/service/service.go
@@ -1992,7 +1992,7 @@ func (process *TeleportProcess) initAuthService() error {
 		// the service has started
 		process.BroadcastEvent(Event{Name: AuthTLSReady, Payload: nil})
 		err := tlsServer.Serve()
-		if err != nil && err != http.ErrServerClosed {
+		if err != nil && errors.Is(err, http.ErrServerClosed) {
 			log.Warningf("TLS server exited with error: %v.", err)
 		}
 		return nil

